python/python/lance/sampler.py

```diff
@@ -0,0 +1,131 @@
+    dataset: lance.LanceDataset,
+    n: int,
+    columns: list[str],
+    batch_size: int,
+    max_takes: int,
+) -> Generator[pa.RecordBatch]:
+    """Sample n records from the dataset.
+
+    Parameters
+    ----------
+    dataset : lance.LanceDataset
+        The dataset to sample from.
+    n : int
+        The number of records to sample.
+    columns : list[str]
+        The columns to load.
+    batch_size : int
+        The batch size to use when loading the data.
+    max_takes : int
+        The maximum number of takes to perform. This is used to limit the number of
+        random reads. Large enough value can give a good random sample without
+        having to issue too much random reads.
+
+    Returns
+    -------
+    Generator of a RecordBatch.
+    """
+    buf: list[pa.RecordBatch] = []
+    total_records = len(dataset)
+    assert total_records > n
+    chunk_size = total_records // max_takes
+    chunk_sample_size = n // max_takes
+    for i in range(0, total_records, chunk_size):
+        offset = i + np.random.randint(0, chunk_size - chunk_sample_size)
+        buf.extend(
+            dataset.to_batches(
+                columns=columns,
+                batch_size=chunk_size,
+                offset=offset,
+                limit=chunk_sample_size,
+            )
+        )
+        if sum([len(b) for b in buf]) >= batch_size:
+            tbl = pa.Table.from_batches(buf)
+            buf.clear()
+            tbl = tbl.combine_chunks()
+            yield tbl.to_batches()[0]
+    if buf:
+        tbl = pa.Table.from_batches(buf).combine_chunks()
+        yield tbl.to_batches()[0]
+
+
+def maybe_sample(
+    dataset: Union[str, Path, lance.LanceDataset],
+    n: int,
+    columns: Union[list[str], str],
+    batch_size: int = 10240,
+    max_takes: int = 8194,
+) -> Generator[pa.RecordBatch]:
+    """Sample n records from the dataset.
+
+    Parameters
+    ----------
+    dataset : Union[str, Path, lance.LanceDataset]
+        The dataset to sample from.
+    n : int
+        The number of records to sample.
+    columns : Union[list[str], str]
+        The columns to load.
+    batch_size : int, optional
+        The batch size to use when loading the data, by default 10240.
+    max_takes : int, optional
+        The maximum number of takes to perform, by default 8194.
+        This is employed to minimize the number of random reads necessary for sampling.
+        A sufficiently large value can provide an effective random sample without
+        the need for excessive random reads.
+
+    Returns
+    -------
+    Generator[pa.RecordBatch]
+        A generator that yields [RecordBatch] of data.
+    """
+    if isinstance(dataset, (str, Path)):
+        dataset = lance.dataset(dataset)
+
+    if isinstance(columns, str):
+        columns = [columns]
+
+    if n >= len(dataset):
+        # Dont have enough data in the dataset. Just do a full scan
+        dataset.to_batches(columns=columns, batch_size=batch_size)
```

Comment:
This should have a return/yeild here?

Reply:
And maybe also add this condition in test_sampler, where. n>len(ds)

Reply:
Good catch. Wrote too much rust recently